File: julia/src/benchmarks/Calculator.jl
using ..BenchmarkFramework

abstract type CalcNode end

struct NumberCalcNode <: CalcNode
    value::Int64
    function NumberCalcNode(value::Int64)
        new(value)
    end
end

struct VariableCalcNode <: CalcNode
    name::String
    function VariableCalcNode(name::String)
        new(name)
    end
end

struct BinaryOpCalcNode <: CalcNode
    op::Char
    left::CalcNode
    right::CalcNode
    function BinaryOpCalcNode(op::Char, left::CalcNode, right::CalcNode)
        new(op, left, right)
    end
end

struct AssignmentCalcNode <: CalcNode
    var::String
    expr::CalcNode
    function AssignmentCalcNode(var::String, expr::CalcNode)
        new(var, expr)
    end
end

mutable struct CalculatorAst <: AbstractBenchmark
    n::Int64
    result::UInt32
    text::String
    expressions::Vector{CalcNode}

    function CalculatorAst()
        n_val = Helper.config_i64("CalculatorAst", "operations")
        new(n_val, UInt32(0), "", CalcNode[])
    end
end

name(b::CalculatorAst)::String = "CalculatorAst"

function generate_random_program(n::Int64 = 1000)::String
    io = IOBuffer()

    write(io, "v0 = 1\n")
    for i in 0:9
        v = i + 1
        write(io, "v$v = v$(v-1) + $v\n")
    end

    for i in 0:n-1
        v = i + 10
        write(io, "v$v = v$(v-1) + ")

        r = Helper.next_int(10)
        if r == 0
            write(io, "(v$(v-1) / 3) * 4 - $i / (3 + (18 - v$(v-2))) % v$(v-3) + 2 * ((9 - v$(v-6)) * (v$(v-5) + 7))")
        elseif r == 1
            write(io, "v$(v-1) + (v$(v-2) + v$(v-3)) * v$(v-4) - (v$(v-5) /  v$(v-6))")
        elseif r == 2
            write(io, "(3789 - (((v$(v-7))))) + 1")
        elseif r == 3
            write(io, "4/2 * (1-3) + v$(v-9)/v$(v-5)")
        elseif r == 4
            write(io, "1+2+3+4+5+6+v$(v-1)")
        elseif r == 5
            write(io, "(99999 / v$(v-3))")
        elseif r == 6
            write(io, "0 + 0 - v$(v-8)")
        elseif r == 7
            write(io, "((((((((((v$(v-6)))))))))) * 2")
        elseif r == 8
            write(io, "$i * (v$(v-1)%6)%7")
        else 
            write(io, "(1)/(0-v$(v-5)) + (v$(v-7))")
        end
        write(io, "\n")
    end

    return String(take!(io))
end

function prepare(b::CalculatorAst)
    b.text = generate_random_program(b.n)
end

mutable struct Parser
    input::String
    pos::Int
    chars::Vector{Char}
    current_char::Char
    expressions::Vector{CalcNode}

    function Parser(input::String)
        chars = collect(input)
        current_char = length(chars) > 0 ? chars[1] : '\0'
        new(input, 1, chars, current_char, CalcNode[])
    end
end

function parse(p::Parser)
    while p.pos <= length(p.chars)
        node = parse_expression(p)
        if node !== nothing
            push!(p.expressions, node)
        end
    end
end

function parse_expression(p::Parser)::CalcNode
    node = parse_term(p)

    while p.pos <= length(p.chars)
        skip_whitespace(p)
        p.pos > length(p.chars) && break

        if p.current_char == '+' || p.current_char == '-'
            op = p.current_char
            advance(p)
            right = parse_term(p)
            node = BinaryOpCalcNode(op, node, right)
        else
            break
        end
    end

    return node
end

function parse_term(p::Parser)::CalcNode
    node = parse_factor(p)

    while p.pos <= length(p.chars)
        skip_whitespace(p)
        p.pos > length(p.chars) && break

        if p.current_char == '*' || p.current_char == '/' || p.current_char == '%'
            op = p.current_char
            advance(p)
            right = parse_factor(p)
            node = BinaryOpCalcNode(op, node, right)
        else
            break
        end
    end

    return node
end

function parse_factor(p::Parser)::CalcNode
    skip_whitespace(p)
    p.pos > length(p.chars) && return NumberCalcNode(0)

    if isdigit(p.current_char)
        return parse_number(p)
    elseif islowercase(p.current_char)
        return parse_variable(p)
    elseif p.current_char == '('
        advance(p)  
        node = parse_expression(p)
        skip_whitespace(p)
        if p.current_char == ')'
            advance(p)  
        end
        return node
    else
        return NumberCalcNode(0)
    end
end

function parse_number(p::Parser)::CalcNode
    v = Int64(0)
    while p.pos <= length(p.chars) && isdigit(p.current_char)
        v = v * 10 + Int64(p.current_char - '0')
        advance(p)
    end
    return NumberCalcNode(v)
end

function parse_variable(p::Parser)::CalcNode
    start = p.pos
    while p.pos <= length(p.chars) && (islowercase(p.current_char) || isdigit(p.current_char))
        advance(p)
    end
    var_name = p.input[start:p.pos-1]

    skip_whitespace(p)
    if p.pos <= length(p.chars) && p.current_char == '='
        advance(p)  
        expr = parse_expression(p)
        return AssignmentCalcNode(var_name, expr)
    end

    return VariableCalcNode(var_name)
end

function advance(p::Parser)
    p.pos += 1
    if p.pos > length(p.chars)
        p.current_char = '\0'
    else
        p.current_char = p.chars[p.pos]
    end
end

function skip_whitespace(p::Parser)
    while p.pos <= length(p.chars) && isspace(p.current_char)
        advance(p)
    end
end

function run(b::CalculatorAst, iteration_id::Int64)
    parser = Parser(b.text)
    parse(parser)
    b.expressions = parser.expressions
    b.result = (b.result + UInt32(length(b.expressions))) & 0xffffffff

    if !isempty(b.expressions)
        last_expr = b.expressions[end]
        if last_expr isa AssignmentCalcNode
            var_name = last_expr.var
            b.result = (b.result + Helper.checksum(var_name)) & 0xffffffff
        end
    end
end

function checksum(b::CalculatorAst)::UInt32
    return b.result
end

mutable struct Interpreter
    variables::Dict{String, Int64}

    function Interpreter()
        new(Dict{String, Int64}())
    end
end

function simple_div(a::Int64, b::Int64)::Int64
    b == 0 && return Int64(0)

    if (a >= 0 && b > 0) || (a < 0 && b < 0)
        div(a, b)  
    else
        -div(abs(a), abs(b))  
    end
end

function simple_mod(a::Int64, b::Int64)::Int64
    b == 0 && return Int64(0)
    a - simple_div(a, b) * b
end

function evaluate(interp::Interpreter, node::CalcNode)::Int64
    if node isa NumberCalcNode
        return node.value
    elseif node isa VariableCalcNode
        return interp.variables[node.name]
    elseif node isa BinaryOpCalcNode
        left = evaluate(interp, node.left)
        right = evaluate(interp, node.right)

        if node.op == '+'
            return left + right
        elseif node.op == '-'
            return left - right
        elseif node.op == '*'
            return left * right
        elseif node.op == '/'
            return simple_div(left, right)
        elseif node.op == '%'
            return simple_mod(left, right)
        else
            return Int64(0)
        end
    elseif node isa AssignmentCalcNode
        value = evaluate(interp, node.expr)
        interp.variables[node.var] = value
        return value
    else
        return Int64(0)
    end
end

function run_interpreter(interp::Interpreter, expressions::Vector{CalcNode})::Int64
    result = Int64(0)
    for expr in expressions
        result = evaluate(interp, expr)
    end
    return result
end

mutable struct CalculatorInterpreter <: AbstractBenchmark
    n::Int64
    ast::Vector{CalcNode}
    result::UInt32

    function CalculatorInterpreter()
        n_val = Helper.config_i64("CalculatorInterpreter", "operations")
        new(n_val, CalcNode[], UInt32(0))
    end
end

name(b::CalculatorInterpreter)::String = "CalculatorInterpreter"

function prepare(b::CalculatorInterpreter)

    calc = CalculatorAst()
    calc.n = b.n
    prepare(calc)
    run(calc, 0)
    b.ast = calc.expressions
end

function run(b::CalculatorInterpreter, iteration_id::Int64)
    interp = Interpreter()
    result = run_interpreter(interp, b.ast)
    b.result += Helper.to_u32(result)
end

function checksum(b::CalculatorInterpreter)::UInt32
    return b.result
end

File: fsharp/Calculator.fs
namespace Benchmarks

open System
open System.Text
open System.Collections.Generic

[<AbstractClass>]
type Node() = class end

type Number(value: int64) =
    inherit Node()
    member _.Value = value

type Variable(name: string) =
    inherit Node()
    member _.Name = name

type BinaryOp(op: char, left: Node, right: Node) =
    inherit Node()
    member _.Op = op
    member _.Left = left
    member _.Right = right

type Assignment(varName: string, expr: Node) =
    inherit Node()
    member _.Var = varName
    member _.Expr = expr

[<AllowNullLiteral>]
type Parser(input: string) =
    let chars = input.ToCharArray()
    let mutable pos = 0
    let mutable currentChar = if chars.Length > 0 then chars.[0] else '\000'
    let expressions = List<Node>()

    let advance() = 
        if pos < chars.Length then
            pos <- pos + 1
            if pos < chars.Length then 
                currentChar <- chars.[pos]
            else
                currentChar <- '\000'

    let skipWhitespace() =
        while pos < chars.Length && Char.IsWhiteSpace(currentChar) do
            advance()

    let rec parseNumber() =
        let mutable value = 0L
        while pos < chars.Length && Char.IsDigit(currentChar) do
            value <- value * 10L + int64(currentChar - '0')
            advance()
        Number(value) :> Node

    let rec parseVariable() =
        let start = pos
        while pos < chars.Length && 
              (Char.IsLetterOrDigit(currentChar) || currentChar = '_') do
            advance()

        let varName = input.Substring(start, pos - start)

        skipWhitespace()
        if pos < chars.Length && currentChar = '=' then
            advance()
            let expr = parseExpression()
            Assignment(varName, expr) :> Node
        else
            Variable(varName) :> Node

    and parseExpression() =
        let mutable node = parseTerm()
        let mutable continueLoop = true

        while pos < chars.Length && continueLoop do
            skipWhitespace()
            if pos >= chars.Length then 
                continueLoop <- false
            else
                let ch = currentChar
                if ch = '+' || ch = '-' then
                    let op = ch
                    advance()
                    let right = parseTerm()
                    node <- BinaryOp(op, node, right) :> Node
                else
                    continueLoop <- false

        node

    and parseTerm() =
        let mutable node = parseFactor()
        let mutable continueLoop = true

        while pos < chars.Length && continueLoop do
            skipWhitespace()
            if pos >= chars.Length then 
                continueLoop <- false
            else
                let ch = currentChar
                if ch = '*' || ch = '/' || ch = '%' then
                    let op = ch
                    advance()
                    let right = parseFactor()
                    node <- BinaryOp(op, node, right) :> Node
                else
                    continueLoop <- false

        node

    and parseFactor() =
        skipWhitespace()
        if pos >= chars.Length then Number(0L) :> Node
        else
            match currentChar with
            | ch when Char.IsDigit(ch) -> parseNumber()
            | ch when Char.IsLetter(ch) -> parseVariable()
            | '(' ->
                advance()
                let node = parseExpression()
                skipWhitespace()
                if pos < chars.Length && currentChar = ')' then advance()
                node
            | _ -> Number(0L) :> Node

    member _.Parse() =
        while pos < chars.Length do
            skipWhitespace()
            if pos >= chars.Length then ()
            else
                let expr = parseExpression()
                expressions.Add(expr)

                skipWhitespace()

        List.ofSeq expressions

type Interpreter() =
    let variables = Dictionary<string, int64>()

    let safeAbs (x: int64) =
        if x >= 0L then x
        elif x = Int64.MinValue then Int64.MaxValue
        else -x

    let simpleDiv (a: int64) (b: int64) =
        if b = 0L then 0L
        elif (a >= 0L && b > 0L) || (a < 0L && b < 0L) then a / b
        else
            let absA = safeAbs a
            let absB = safeAbs b
            -(absA / absB)

    let simpleMod (a: int64) (b: int64) =
        if b = 0L then 0L
        else a - simpleDiv a b * b

    let rec evaluate (node: Node) =
        match node with
        | :? Number as num -> num.Value
        | :? Variable as var ->
            match variables.TryGetValue(var.Name) with
            | true, value -> value
            | false, _ -> 0L
        | :? BinaryOp as bin ->
            let left = evaluate bin.Left
            let right = evaluate bin.Right

            match bin.Op with
            | '+' -> left + right
            | '-' -> left - right
            | '*' -> left * right
            | '/' -> simpleDiv left right
            | '%' -> simpleMod left right
            | _ -> 0L
        | :? Assignment as assign ->
            let value = evaluate assign.Expr
            variables.[assign.Var] <- value
            value
        | _ -> 0L

    member _.Run(expressions: Node list) =
        let mutable result = 0L
        for expr in expressions do
            result <- evaluate expr
        result

module ProgramGenerator =
    let generateRandomProgram (n: int64) =
        let sb = StringBuilder()
        sb.AppendLine("v0 = 1") |> ignore

        for i = 0 to 9 do
            let v = i + 1
            sb.AppendLine($"v{v} = v{v - 1} + {v}") |> ignore

        let mutable i = 0L
        while i < n do
            let v = int(i + 10L)
            sb.Append($"v{v} = v{v - 1} + ") |> ignore

            match Helper.NextInt(10) with
            | 0 -> sb.AppendFormat("(v{0} / 3) * 4 - {1} / (3 + (18 - v{2})) % v{3} + 2 * ((9 - v{4}) * (v{5} + 7))", 
                    v - 1, i, v - 2, v - 3, v - 6, v - 5) |> ignore
            | 1 -> sb.AppendFormat("v{0} + (v{1} + v{2}) * v{3} - (v{4} / v{5})", 
                    v - 1, v - 2, v - 3, v - 4, v - 5, v - 6) |> ignore
            | 2 -> sb.AppendFormat("(3789 - (((v{0})))) + 1", v - 7) |> ignore
            | 3 -> sb.AppendFormat("4/2 * (1-3) + v{0}/v{1}", v - 9, v - 5) |> ignore
            | 4 -> sb.AppendFormat("1+2+3+4+5+6+v{0}", v - 1) |> ignore
            | 5 -> sb.AppendFormat("(99999 / v{0})", v - 3) |> ignore
            | 6 -> sb.AppendFormat("0 + 0 - v{0}", v - 8) |> ignore
            | 7 -> sb.AppendFormat("((((((((((v{0})))))))))) * 2", v - 6) |> ignore
            | 8 -> sb.AppendFormat("{0} * (v{1}%6)%7", i, v - 1) |> ignore
            | 9 -> sb.AppendFormat("(1)/(0-v{0}) + (v{1})", v - 5, v - 7) |> ignore
            | _ -> ()

            sb.AppendLine() |> ignore
            i <- i + 1L

        sb.ToString()

type CalculatorAst() =
    inherit Benchmark()

    let mutable n = 0L
    let mutable text = ""
    let mutable result = 0u
    let mutable expressions: Node list = []

    member this.SetOperations(operations: int64) =
        n <- operations
        text <- ProgramGenerator.generateRandomProgram n

    member this.Parse() =
        let parser = Parser(text)
        expressions <- parser.Parse()

        result <- result + uint32 expressions.Length

        match expressions with
        | [] -> ()
        | _ ->
            let last = expressions |> List.last
            match last with
            | :? Assignment as assign -> result <- result + Helper.Checksum(assign.Var)
            | _ -> ()

    member this.Expressions = expressions

    override this.Checksum = result
    override this.Name = "Calculator::Ast"

    override this.Prepare() =
        n <- this.ConfigVal("operations")
        this.SetOperations(n)
        result <- 0u  

    override this.Run(_: int64) =
        this.Parse()  

type CalculatorInterpreter() =
    inherit Benchmark()

    let mutable n = 0L
    let mutable result = 0u
    let mutable ast: Node list = []

    override this.Checksum = result
    override this.Name = "Calculator::Interpreter"

    override this.Prepare() =
        n <- this.ConfigVal("operations")

        let calculator = CalculatorAst()
        calculator.SetOperations(n)
        calculator.Parse()

        ast <- calculator.Expressions

        result <- 0u

    override this.Run(_: int64) =
        let interpreter = Interpreter()
        let calcResult = interpreter.Run(ast)
        result <- result + uint32 calcResult